overlays/README.md

```diff
@@ -43,3 +43,70 @@ Dependencies are in a requirements file. To run tests:
+
+Our environment stack represents a DAG of dependencies. With unsaved files, there are now two DAGs - one corresponding to the saved contents and one corresponding to the unsaved contents.
+
+Proposal: Have two cache tables per environment. The first one is the existing cache table, containing all saved modules and their values. The second, new one just contains information about unsaved modules.
+
+When pushing an unsaved file update, we filter dependencies so that only unsaved modules are updated. When pushing a saved file update, we propagate changes to all dependencies.
+
+The above scheme means that an unsaved module will be represented twice, once in the saved table and once in the unsaved table. We need this because each environment operation has two variants: one that works on the original DAG and the other that works on the new DAG.
+
+When getting a value that is not currently in cache: If the key is a saved module, we need to use the original environment (*even if* we depend on a key that is an unsaved module). Otherwise, we would end up making saved module values depend on unsaved module values. If the key is an unsaved module, we need to use the wrapper environment.
+
+For example, suppose computing a value for an unsaved module `b` depends on a saved module class `a.X`, but `a.X` is uncached. In this case, we have to compute the value for `a.X` ... but in the original DAG. This is because we do not want `a.X` to use the unsaved values of any of its dependents. So, we may need to compute `b.XParent` - the parent of `a.X` - using the saved contents of `b`, even though `b` is unsaved. So, we need to keep both results around.
+
+There is no need to wrap the cache table or do any pass-through. We will automatically look up or write to the correct table based on what the request is - whether the `get` call expects you to use the saved or unsaved values of your dependents and whether the `update` push call is
+
+This design makes it easy to go for any of the other approaches - for example, if we want to recompute from scratch without persisting, we can disable the persistent cache table.
+
+Note: This reuses the same push-based approach as regular file updates. It can thus easily invalidate old cache values when updating, unlike the previous overlay approach.
+
+Note: The type and size of the extra cache needed is the same in both cases. So, if we would go with an in-memory cache for the persistent-overlay approach, we would do the same for persistent. If we would go with a pre-allocated shared memory table, we would do the same here.
+
+Question: Have one wrapped environment for each unsaved file or one wrapped environment for all unsaved files?
+
+Actually, once we have caching, we cannot have multiple environments. Otherwise, if we update the original environment independently, there would be nothing to tell the overlay environment to invalidate its cache.
```

Comment:
Fair enough. I didn't want to keep multiple such tables around, since then we'd have to think of some way to allocate N tables or give them unique keys within the same table. With a single unsaved hash table, we can just pre-allocate `MAX_UNSAVED_MODULES * MAX_SIZE_PER_MODULE`. But no strong feelings here - whatever is easier.